File: frontend/src/types/billing.ts
export type PaymentStatus = "Pending" | "Partially Paid" | "Paid" | "Cancelled";

export interface BillingRecord {
  id: number;
  case_id: number;
  case_no: string;
  applicant: string | null;
  bank: string | null;
  city: string | null;
  executive: string | null;
  bank_payout_amount: string;
  bank_paid_amount: string;
  bank_payment_status: PaymentStatus;
  bank_paid_date: string | null;
  bank_payment_reference: string | null;
  executive_payout_amount: string;
  executive_paid_amount: string;
  executive_payment_status: PaymentStatus;
  executive_paid_date: string | null;
  executive_payment_reference: string | null;
  gross_margin: string;
  bank_balance: string;
  executive_balance: string;
  expected_gross_margin: string;
  realized_cash_margin: string;
  remarks: string | null;
  created_at: string;
  updated_at: string;
}

export interface BillingPayload {
  case_id: number;
  bank_payout_amount: string;
  bank_paid_amount: string;
  bank_payment_status: PaymentStatus;
  bank_paid_date: string | null;
  bank_payment_reference: string | null;
  executive_payout_amount: string;
  executive_paid_amount: string;
  executive_payment_status: PaymentStatus;
  executive_paid_date: string | null;
  executive_payment_reference: string | null;
  remarks: string | null;
}

export interface BillingFilters {
  case_no?: string;
  bank?: string;
  executive?: string;
  city?: string;
  bank_payment_status?: PaymentStatus;
  executive_payment_status?: PaymentStatus;
  from_date?: string;
  to_date?: string;
}

export interface BulkPreviewFilters {
  case_ids?: number[]; receive_date_from?: string; receive_date_to?: string; bank?: string;
  city?: string; executive?: string; status?: string; only_without_billing?: boolean;
}
export type RateStatus = "MATCHED" | "MISSING" | "AMBIGUOUS";
export interface BulkPreviewRow {
  case_id: number; case_no: string; applicant: string | null; bank: string | null; city: string | null;
  location: string | null; executive: string | null; loan_type: string | null; product_type: string | null;
  bank_rate_status: RateStatus; bank_rate_id: number | null; bank_payout_amount: string | null;
  executive_rate_status: RateStatus; executive_rate_id: number | null; executive_payout_amount: string | null;
  expected_gross_margin: string | null; existing_billing: boolean; validation_errors: string[]; ready: boolean;
}
export interface BulkPreviewResponse {
  rows: BulkPreviewRow[];
  summary: { selected_cases: number; ready_cases: number; missing_bank_rates: number; missing_executive_rates: number; ambiguous_rates: number; existing_billing_records: number };
}
export interface BulkCreateResponse {
  created_count: number; skipped_count: number; error_count: number;
  results: { case_id: number; case_no: string | null; status: "CREATED" | "SKIPPED" | "ERROR"; billing_id: number | null; errors: string[] }[];
}

File: backend/app/schemas/billing.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


PaymentStatus = Literal["Pending", "Partially Paid", "Paid", "Cancelled"]


class BillingBase(BaseModel):
    bank_payout_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    bank_paid_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    bank_payment_status: PaymentStatus = "Pending"
    bank_paid_date: date | None = None
    bank_payment_reference: str | None = None
    executive_payout_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    executive_paid_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    executive_payment_status: PaymentStatus = "Pending"
    executive_paid_date: date | None = None
    executive_payment_reference: str | None = None
    remarks: str | None = None


class BillingCreate(BillingBase):
    case_id: int


class BillingUpdate(BaseModel):
    case_id: int | None = None
    bank_payout_amount: Decimal | None = Field(default=None, ge=0)
    bank_paid_amount: Decimal | None = Field(default=None, ge=0)
    bank_payment_status: PaymentStatus | None = None
    bank_paid_date: date | None = None
    bank_payment_reference: str | None = None
    executive_payout_amount: Decimal | None = Field(default=None, ge=0)
    executive_paid_amount: Decimal | None = Field(default=None, ge=0)
    executive_payment_status: PaymentStatus | None = None
    executive_paid_date: date | None = None
    executive_payment_reference: str | None = None
    remarks: str | None = None

    @model_validator(mode="after")
    def reject_null_required_fields(self):
        required_when_supplied = {
            "case_id",
            "bank_payout_amount",
            "bank_paid_amount",
            "bank_payment_status",
            "executive_payout_amount",
            "executive_paid_amount",
            "executive_payment_status",
        }
        for field in required_when_supplied & self.model_fields_set:
            if getattr(self, field) is None:
                raise ValueError(f"{field} cannot be null")
        return self


class BillingResponse(BillingBase):
    id: int
    case_id: int
    case_no: str
    los_no: str | None = None
    applicant: str | None = None
    bank: str | None = None
    city: str | None = None
    executive: str | None = None
    gross_margin: Decimal
    bank_balance: Decimal
    executive_balance: Decimal
    expected_gross_margin: Decimal
    realized_cash_margin: Decimal
    created_at: datetime
    updated_at: datetime
    bank_payout_rate_id: int | None = None
    executive_payout_rate_id: int | None = None

    model_config = ConfigDict(from_attributes=True)


class BulkBillingRequest(BaseModel):
    case_ids: list[int] | None = Field(default=None, max_length=5000)
    receive_date_from: date | None = None
    receive_date_to: date | None = None
    bank: str | None = None
    city: str | None = None
    executive: str | None = None
    status: str | None = None
    only_without_billing: bool = True

    @model_validator(mode="after")
    def validate_selection(self):
        if self.case_ids is not None and not self.case_ids:
            raise ValueError("case_ids cannot be empty")
        if self.receive_date_from and self.receive_date_to and self.receive_date_to < self.receive_date_from:
            raise ValueError("receive_date_to must be on or after receive_date_from")
        return self


class BulkPreviewRow(BaseModel):
    case_id: int
    case_no: str
    los_no: str | None = None
    applicant: str | None
    bank: str | None
    city: str | None
    location: str | None = None
    executive: str | None
    loan_type: str | None
    product_type: str | None
    bank_rate_status: Literal["MATCHED", "MISSING", "AMBIGUOUS"]
    bank_rate_id: int | None
    bank_payout_amount: Decimal | None
    executive_rate_status: Literal["MATCHED", "MISSING", "AMBIGUOUS"]
    executive_rate_id: int | None
    executive_payout_amount: Decimal | None
    expected_gross_margin: Decimal | None
    existing_billing: bool
    validation_errors: list[str]
    ready: bool


class BulkPreviewSummary(BaseModel):
    selected_cases: int
    ready_cases: int
    missing_bank_rates: int
    missing_executive_rates: int
    ambiguous_rates: int
    existing_billing_records: int


class BulkPreviewResponse(BaseModel):
    rows: list[BulkPreviewRow]
    summary: BulkPreviewSummary


class BulkCreateRequest(BaseModel):
    case_ids: list[int] = Field(min_length=1, max_length=5000)


class BulkCreateResult(BaseModel):
    case_id: int
    case_no: str | None = None
    los_no: str | None = None
    status: Literal["CREATED", "SKIPPED", "ERROR"]
    billing_id: int | None = None
    errors: list[str] = Field(default_factory=list)


class BulkCreateResponse(BaseModel):
    created_count: int
    skipped_count: int
    error_count: int
    results: list[BulkCreateResult]
